File: 2026-bogota-agent-sdk/typescript/07_subagente.ts
// 07 · Un subagente: otro agente con su propio contexto, sus tools y su prompt.
// El principal no tiene acceso al correo. Delega, y solo recibe el resumen.
// Cada mensaje que viene de adentro trae parent_tool_use_id con el id de la delegación.
import "./entorno.ts"; // carga ANTHROPIC_API_KEY de .env
import { query } from "@anthropic-ai/claude-agent-sdk";
import { servidorGmail } from "./gmail_mcp.ts";

const conversacion = query({
  prompt:
    "Hoy es 19 de septiembre de 2026. Averigua cuándo es mi próximo vuelo delegando " +
    "en el subagente lector-de-correos, con run_in_background en false, y espera su " +
    "respuesta antes de contestarme.",
  options: {
    model: "claude-sonnet-5",
    systemPrompt:
      "Eres un asistente personal. Respondes en español, corto y con datos. " +
      "Usa tus tools y skills; no adivines fechas ni reservas.",
    tools: ["Task"], // el principal solo puede delegar (en el init la tool se llama Task)
    mcpServers: { gmail: servidorGmail },
    // Los permisos NO se heredan del bloque agents: van aquí, en el permiso global.
    allowedTools: ["mcp__gmail__buscar", "mcp__gmail__leer"],

    agents: {
      "lector-de-correos": {
        description: "Lee el correo del usuario y devuelve solo los datos duros del vuelo.",
        prompt:
          "Eres un lector de correos y respondes en español. Busca desde el 12 de " +
          "septiembre de 2026, lee los " +
          "candidatos e ignora publicidad, encuestas y hoteles. Devuelve en dos líneas: " +
          "ruta, día, hora y código de reserva del vuelo futuro más cercano.",
        tools: ["mcp__gmail__buscar", "mcp__gmail__leer"],
        mcpServers: ["gmail"], // hereda el servidor por nombre
        model: "sonnet",
        background: false, // el principal espera la respuesta en vez de seguir
        maxTurns: 10,
      },
    },

    forwardSubagentText: true, // sin esto no se ve lo que dice el subagente
    settingSources: [],
    strictMcpConfig: true,
    maxTurns: 12,
    maxBudgetUsd: 0.20,
    env: { ...process.env, CLAUDE_CODE_DISABLE_AUTO_MEMORY: "1" },
  },
});

for await (const m of conversacion) {
  if (m.type === "assistant") {
    const quien = m.parent_tool_use_id ? `  [subagente ${m.parent_tool_use_id.slice(-6)}]` : "[principal]";
    for (const b of m.message.content) {
      if (b.type === "tool_use") console.log(`${quien} -> ${b.name}`);
      if (b.type === "text" && b.text.trim()) console.log(`${quien} ${b.text.trim()}`);
    }
  }
  if (m.type === "result" && m.subtype === "success") {
    console.log(`\nCosto total (incluye el subagente): $${m.total_cost_usd.toFixed(4)}`);
  }
}

File: 2026-bogota-agent-sdk/python/07_subagente.py
"""07 · un subagente — delegar para no ensuciar el contexto.

`lector-de-correos` busca y lee en la bandeja, y devuelve UNA frase. Su contexto
es propio: el padre nunca ve los 4 correos completos, solo el resumen. Eso es lo
que compra un subagente, además de poder correr varios en paralelo.

Lo único que viaja del padre al hijo es el prompt de la llamada a la tool `Agent`.
El hijo no ve la conversación del padre.

Dos detalles que confunden en vivo:
  - la tool aparece como "Agent" en el stream y como "Task" en la lista del ARRANQUE
  - los mensajes del hijo traen `parent_tool_use_id`; el impresor los marca [subagente]

    uv run python 07_subagente.py
"""

import entorno  # carga ANTHROPIC_API_KEY de .env

import asyncio
import sys
from pathlib import Path

from claude_agent_sdk import (
    AgentDefinition, ClaudeAgentOptions, create_sdk_mcp_server, query, tool,
)

from mostrar import linea

AQUI = Path(__file__).parent.resolve()
PREGUNTA = "Revisa mis correos de la última semana y dime cuándo es mi próximo vuelo."

SYSTEM_PROMPT = """Eres un asistente personal. Respondes en español, corto y con datos.
Usa tus tools y skills; no adivines fechas ni reservas."""


@tool("hoy", "La fecha de hoy. Pídela siempre antes de razonar sobre fechas.", {})
async def hoy(args):
    return {"content": [{"type": "text", "text": "sábado 19 de septiembre de 2026"}]}


RELOJ = create_sdk_mcp_server(name="reloj", version="1.0.0", tools=[hoy])

LECTOR = AgentDefinition(
    description="Lee la bandeja de correo y devuelve los vuelos que encuentre. "
                "Úsalo cuando haya que revisar correos.",
    prompt="Haz UNA sola búsqueda con el rango que te pidan y lee los correos que "
           "puedan ser vuelos. Devuelve una línea por vuelo: fecha, hora, ruta y "
           "código de reserva. No opines ni expliques el proceso.",
    tools=["mcp__gmail__buscar", "mcp__gmail__leer"],   # lo que no está aquí, no existe para él
    model="haiku",                                      # el hijo puede ser más barato que el padre
    maxTurns=8,                     # en Python los campos de varias palabras van en camelCase
    background=False,               # en primer plano, para verlo en el stream
)


async def main():
    opciones = ClaudeAgentOptions(
        model="claude-sonnet-5",
        system_prompt=SYSTEM_PROMPT,
        agents={"lector-de-correos": LECTOR},
        tools=["Agent"],
        mcp_servers={
            "reloj": RELOJ,
            "gmail": {"command": sys.executable, "args": [str(AQUI / "gmail_mcp.py")]},
        },
        allowed_tools=["Agent", "mcp__reloj__hoy", "mcp__gmail__buscar", "mcp__gmail__leer"],
        forward_subagent_text=True,        # sin esto solo verías sus tool calls
        setting_sources=[],
        strict_mcp_config=True,
        env={"CLAUDE_CODE_DISABLE_AUTO_MEMORY": "1"},
        max_turns=12,
        max_budget_usd=0.25,
        stderr=lambda _: None,
    )

    print(f"PREGUNTA  {PREGUNTA}\n")
    async for m in query(prompt=PREGUNTA, options=opciones):
        linea(m)


asyncio.run(main())
